src/sql/src/pure.rs

```diff
@@ -211,6 +218,134 @@ pub async fn purify_create_source(
                 .await?;
             let tables = mz_postgres_util::publication_info(&config, &publication).await?;
 
+            let mut targeted_subsources = vec![];
+
+            let mut validated_requested_subsources = vec![];
+            match requested_subsources {
+                Some(CreateSourceSubsources::All) => {
+                    for table in &tables {
+                        let name =
+                            UnresolvedObjectName::qualified(&[&table.namespace, &table.name]);
+                        validated_requested_subsources.push((name.clone(), name, table));
+                    }
+                }
+                Some(CreateSourceSubsources::Subset(subsources)) => {
+                    // The user manually selected a subset of upstream tables so we need to
+                    // validate that the names actually exist and are not ambiguous
+
+                    // An index from table_name -> schema_name -> PostgresTableDesc
+                    let mut tables_by_name = HashMap::new();
+                    for table in &tables {
+                        tables_by_name
+                            .entry(table.name.clone())
+                            .or_insert_with(HashMap::new)
+                            .entry(table.namespace.clone())
+                            .or_insert(table);
+                    }
+
+                    for subsource in subsources {
+                        let (upstream_name, subsource_name) = match subsource.clone() {
+                            CreateSourceSubsource::Bare(name) => (name.clone(), name),
+                            CreateSourceSubsource::Aliased(name, alias) => (name, alias),
+                            CreateSourceSubsource::Resolved(_, _) => {
+                                bail!("Cannot alias subsource using `INTO`, use `AS` instead")
```

Comment:
Yes. There is another such match in the planner but it's only there because I had to handle the match arm. This is where the actual error will be produced since purification runs before planning